ports/stm32/storage.c

```diff
@@ -233,41 +238,194 @@ mp_uint_t storage_write_blocks(const uint8_t *src, uint32_t block_num, uint32_t
+STATIC mp_obj_t pyb_flash_writeblocks(size_t n_args, const mp_obj_t *args) {
+    pyb_flash_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    uint32_t block_num = mp_obj_get_int(args[1]);
     mp_buffer_info_t bufinfo;
-    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
-    mp_uint_t ret = storage_write_blocks(bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / FLASH_BLOCK_SIZE);
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_READ);
+    mp_uint_t ret;
+    if (n_args == 3) {
+        // Cast self->start to signed in case it's pyb_flash_obj with negative start
+        block_num += FLASH_PART1_START_BLOCK + (int32_t)self->start / FLASH_BLOCK_SIZE;
+        ret = storage_write_blocks(bufinfo.buf, block_num, bufinfo.len / FLASH_BLOCK_SIZE);
+    } else {
+        uint32_t offset = mp_obj_get_int(args[3]);
+        #if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE
+        block_num += self->start / FLASH_BLOCK_SIZE;
+        ret = flash_bdev_writeblocks_ext(bufinfo.buf, block_num, offset, bufinfo.len);
+        #elif defined(SPIFLASH)
+        block_num += self->start / MP_SPIFLASH_ERASE_BLOCK_SIZE;
+        ret = spi_bdev_writeblocks_raw(SPIFLASH, bufinfo.buf, block_num, offset, bufinfo.len);
+        #else
+        ret = -MP_EIO;
+        #endif
+    }
     return MP_OBJ_NEW_SMALL_INT(ret);
 }
-STATIC MP_DEFINE_CONST_FUN_OBJ_3(pyb_flash_writeblocks_obj, pyb_flash_writeblocks);
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(pyb_flash_writeblocks_obj, 3, 4, pyb_flash_writeblocks);
 
-STATIC mp_obj_t pyb_flash_ioctl(mp_obj_t self, mp_obj_t cmd_in, mp_obj_t arg_in) {
+STATIC mp_obj_t pyb_flash_ioctl(mp_obj_t self_in, mp_obj_t cmd_in, mp_obj_t arg_in) {
+    pyb_flash_obj_t *self = self_in;
     mp_int_t cmd = mp_obj_get_int(cmd_in);
     switch (cmd) {
-        case MP_BLOCKDEV_IOCTL_INIT: storage_init(); return MP_OBJ_NEW_SMALL_INT(0);
+        case MP_BLOCKDEV_IOCTL_INIT: {
+            mp_int_t ret = 0;
+            storage_init();
+            if (mp_obj_get_int(arg_in) == 1) {
+                // Will be using extended block protocol
+                if (self == &pyb_flash_obj) {
+                    ret = -1;
+                #if defined(SPIFLASH)
+                } else {
+                    self->use_spi_raw = true;
+                #endif
+                }
+            }
+            return MP_OBJ_NEW_SMALL_INT(ret);
+        }
         case MP_BLOCKDEV_IOCTL_DEINIT: storage_flush(); return MP_OBJ_NEW_SMALL_INT(0); // TODO properly
         case MP_BLOCKDEV_IOCTL_SYNC: storage_flush(); return MP_OBJ_NEW_SMALL_INT(0);
-        case MP_BLOCKDEV_IOCTL_BLOCK_COUNT: return MP_OBJ_NEW_SMALL_INT(storage_get_block_count());
-        case MP_BLOCKDEV_IOCTL_BLOCK_SIZE: return MP_OBJ_NEW_SMALL_INT(storage_get_block_size());
+
+        case MP_BLOCKDEV_IOCTL_BLOCK_COUNT: {
+            mp_int_t n;
+            if (self == &pyb_flash_obj) {
+                // Get true size
+                n = storage_get_block_count();
```

Comment:
It would be worth adding a comment where `pyb_flash_obj` is defined explaining that it has self->len set to zero (because it's explicitly handled here).

Reply:
done